src/mds/MDSContext.cc

```diff
@@ -28,6 +28,9 @@ void MDSInternalContextBase::complete(int r) {
   assert(mds != NULL);
   assert(mds->mds_lock.is_locked_by_me());
   MDSContext::complete(r);
+
+  // avoid unhealthy heartbeat map when finish_contexts() processes large context list
+  mds->heartbeat_reset();
```

Comment:
Large context list are delayed messages (caused by frozen subtree). how large it is depends on how many messages were sent by clients while subtree was frozen. these messages need to be processed before dispatching new messages (to guarantee message order).  mds can't directly control how large the context list is (mds can control it indirectly by reducing max_export_size) . I can't figure out a better approach.

Reply:
Thanks for explaining. What concerns me is that this hides the problem: the MDS is stuck processing a large context list and not doing anything else. This is precisely the type of scenario the heartbeat mechanism is designed to catch!

How about reducing the `max_export_size`? `1G` and even `100M` sounds too large. What are the drawbacks to breaking the exports into smaller chunks like `10M`?

Reply:
10M is about 5000 inodes, that's too small

Reply:
10M is about 5000 inodes, that's too small.  besides, we need to change the code that chooses export candidate. it's too inefficiency to freeze a large subtree, but only export small portion of it.

Reply:
large number of delayed messages is a generic problem. mds also may delay large number of messages during recovering. I don't think we can have better option for current code base (with the large mds_mutex)